### PR TITLE
Research techs only available to each civ

### PR DESCRIPTION
**Pull request by andy5995**
For
https://wildfiregames.com/forum/topic/86883-final-era-full-upgrades/?do=findComment&comment=511794

### CODE REVIEW

File: maps/scripts/DeathMatch.js [hunk 1]
```diff
@@ -15,7 +15,12 @@
         if (!cmpTechnologyManager)
             continue;
 
-        for (const templateName of TechnologyTemplates.GetNames().filter(a => TechnologyTemplates.Has(a)))
+        for (const templateName of TechnologyTemplates.GetNames().filter(a => TechnologyTemplates.Has(a))) {
+
+            if ((templateName == "phase_city_generic" || templateName == "phase_town_generic") && (civ == "athen" || civ == "pers")
```
Comment (andy5995) [inline maps/scripts/DeathMatch.js]: @StanleySweet How do I look up the civ of the player?

**Comment by andy5995** [comment]
@StanleySweet Would this do what you want?

File: maps/scripts/DeathMatch.js [hunk 2]
```diff
@@ -15,7 +15,13 @@
         if (!cmpTechnologyManager)
             continue;
 
-        for (const templateName of TechnologyTemplates.GetNames().filter(a => TechnologyTemplates.Has(a)))
+        const civ = Engine.QueryInterface(cmpPlayerManager.GetPlayerByID(i), IID_Identity).GetCiv();
+        for (const templateName of TechnologyTemplates.GetNames().filter(a => TechnologyTemplates.Has(a))) {
+
+            if ((templateName === "phase_city" || templateName === "phase_town") && (civ === "athen" || civ === "pers"))
```
Comment (andy5995) [inline maps/scripts/DeathMatch.js]: Hmmm... well this doesn't work. They actually get an extra City upgrade somehow, lol.

**Comment by JustusAvramenko** [comment]
Also, I don't think the civs should be hard coded like that. It should be genericized, methinks.

**Comment by StanleySweet** [comment]
Might be complicated to do. I don't know if we have a access to which phases are being overriden there.

It's the same issue for checkrefs by the way.

**Comment by andy5995** [comment]
Instead of researching all techs, is there a way to look at each structure template for a civ and research techs referenced in them?

**Comment by StanleySweet** [comment]
Yep. You open each template that exists, then look at its identity component to see if it matches the civ of the current player. Then you look at its researcher component

**Comment by andy5995** [comment]
> Yep. You open each template that exists, 

@StanleySweet Could you please give me a link to a code example?

**Comment by Freagarach** [comment]
https://code.wildfiregames.com/source/0ad/browse/ps/trunk/binaries/data/mods/public/simulation/components/AIInterface.js$221

**Comment by andy5995** [comment]
Thanks @Freagarach!

> Then you look at its researcher component

Could you also give me a link to a code example for this?

**Comment by StanleySweet** [comment]
Just template.Researcher I think

**Comment by andy5995** [comment]
> Just template.Researcher I think

That returns a string like

```
({Technologies:{'@datatype':"tokens", _string:"trader_health trade_gain_01 trade_gain_02 trade_commercial_treaty"}})
```

But when I try to use `template.Researcher.GetTechnologiesList()` I get an error saying that it's not a function.

Example code

```
    let cmpTemplateManager = Engine.QueryInterface(SYSTEM_ENTITY, IID_TemplateManager);
    let techs = cmpTemplateManager.GetTemplateWithoutValidation("structures/" + civ + "/market").Researcher;
    warn(uneval(techs.GetTechnologiesList()));
```

How do I get just an array of the techs?

**Comment by andy5995** [comment]
> How do I get just an array of the techs?

This works:

```
    let cmpTemplateManager = Engine.QueryInterface(SYSTEM_ENTITY, IID_TemplateManager);
    let techs = cmpTemplateManager.GetTemplateWithoutValidation("structures/" + civ + "/market").Researcher.Technologies._string.split(" ");
    warn(uneval(techs));
```

**Comment by StanleySweet** [comment]
Sweet work.

**Comment by andy5995** [comment]
I just found that this still isn't working as intended. It will research more techs than it should. For example, it will research "gather_farming_seed_drill" for any civ, not just Han (even though it should only be available to Han).

from `gather_farming_seed_drill.json`

```
	"requirements": {
		"all": [
			{ "tech": "phase_village" },
			{ "civ": "han" }
		]
	},
```

**Comment by StanleySweet** [comment]
You already pass template.top, I assume it shouldn't be too hard to check for 


```js
template.requirements.some(r => {
//TODO: Check for r.all
//TODO: Check for r.any
//TODO: Check for r.notciv  
//TODO: Check for cases I might be missing by looking for requirements in simulation/data/technologies. You could also probably steal the code that does that.
if ( r.civ === civ )
return true;


return false;
});
```

**Comment by StanleySweet** [comment]
I don't remember, does this work now?

**Comment by andy5995** [comment]
> I don't remember, does this work now?

I just pushed a new commit. It should work a lot better now but hasn't been fully tested. I started a game with gauls and han. I confirmed han got the seed_drill upgrade and other civs didn't. Gauls get harvesting and other civs don't, etc.

**Pull request merged**